server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -872,20 +878,40 @@ public IndexMetadata withInSyncAllocationIds(int shardId, Set<String> inSyncSet)
     }
 
     /**
-     * Creates a copy of this instance that has the primary term for the given shard id incremented.
-     * @param shardId shard id to increment primary term for
-     * @return updated instance with incremented primary term
+     * Update this instance for a new node allocation.
+     * @param shardId shard id the allocation is for
+     * @param increaseTerm {@code true} if the primary term should be increased
+     * @param newNodeIndexVersion the index version of the new node
+     * @return instance suitable updated
      */
-    public IndexMetadata withIncrementedPrimaryTerm(int shardId) {
-        final long[] incremented = this.primaryTerms.clone();
-        incremented[shardId]++;
+    public IndexMetadata updateForAllocation(int shardId, boolean increaseTerm, IndexVersion newNodeIndexVersion) {
+        boolean modified = false;
+        long[] primaryTerms = this.primaryTerms;
+        if (increaseTerm) {
+            primaryTerms = primaryTerms.clone();
+            primaryTerms[shardId]++;
+            modified = true;
+        }
+        IndexVersion shardIndexVersion = shardReadVersions.getOrDefault(shardId, IndexVersions.ZERO);
+        Map<Integer, IndexVersion> shardReadVersions = this.shardReadVersions;
+        if (newNodeIndexVersion.after(shardIndexVersion)) {
+            shardReadVersions = new HashMap<>(shardReadVersions);
+            shardReadVersions.put(shardId, newNodeIndexVersion);
+            shardReadVersions = Map.copyOf(shardReadVersions);
+            modified = true;
+        }
```

Comment:
shardId is always a number in [0, getNumberOfShards()). It might be easier to use list/array here instead of the map, similarly to: https://github.com/elastic/elasticsearch/pull/90362